---
title: "Gene enrichment for subclusters cluster-0 (cell type: Oligodendrocyte) of RD_VA_01-12 integrated data"
author: "K4"
date: "07/10/2023"
output:
  html_document:
    df_print: paged
---

```{r, results=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
# Markdown parameters
set_echo = FALSE
set_eval = FALSE
```

```{r, results=FALSE, echo=FALSE, message=FALSE, warning=FALSE}
# Function to set markdown text color
colorize <- function(x, color) {
  if (knitr::is_latex_output()) {
    sprintf("\\textcolor{%s}{%s}", color, x)
  } else if (knitr::is_html_output()) {
    sprintf("<span style='color: %s;'>%s</span>", color,
      x)
  } else x
}
```


```{r, include=FALSE}
# Set height of chunk output
options(width = 60)
local({
  hook_output <- knitr::knit_hooks$get('output')
  knitr::knit_hooks$set(output = function(x, options) {
    if (!is.null(options$max.height)) options$attr.output <- c(
      options$attr.output,
      sprintf('style="max-height: %s;"', options$max.height)
    )
    hook_output(x, options)
  })
})
```


```{r, results = FALSE, echo=set_echo, message=FALSE, warning=FALSE}
# Load packages
library(dplyr)
library(Seurat)
library(patchwork)
library(ggplot2)
library(cowplot)
library(rmarkdown)
```

<!-- Information of Used Library -->
### `r colorize('Information of Used R Libraries', 'blue')` 

The following R libraries were used to process the data and produce this report.

dplyr: `r packageVersion('dplyr')`, 

Seurat: `r packageVersion('Seurat')`, 

patchwork: `r packageVersion('patchwork')`, 

ggplot2: `r packageVersion('ggplot2')`, 

cowplot: `r packageVersion('cowplot')`.


```{r, echo=FALSE}
# Useful string constant
suffix <- "Seurat"
sampleNames_filt <- c("RD_VA_01", "RD_VA_02", "RD_VA_03", "RD_VA_05",  
                 "RD_VA_07", "RD_VA_08", "RD_VA_09", "RD_VA_10","RD_VA_11","RD_VA_12")
sample_Names_10 = gsub("_", "-", sampleNames_filt)
rawDataDir_1 <- "/projectnb/czlab/A10_Rosene/Data/20230408_combinedTwo_results/"
rawDataDir_2 <- "/outs/filtered_feature_bc_matrix"
projectName <- "RD_VA_01_12"

# Seurat parameters
variable.features <- 2000
nPCs <- 50
```


```{r, echo=FALSE}
# Load subclustering data
data.combined <- readRDS("data_Integrated_10_samples_5_celltypes.RDS")
```




<!-- # ```{r, echo=FALSE} -->
<!-- # # Load subclustering data -->
<!-- # sets = data.combined$group -->
<!-- # sets[sets == "RD-VA-01"] <- "CR" -->
<!-- # sets[sets == "RD-VA-08"] <- "CR" -->
<!-- # sets[sets == "RD-VA-11"] <- "CR" -->
<!-- # sets[sets == "RD-VA-12"] <- "CR" -->
<!-- #  -->
<!-- # sets[sets == "RD-VA-02"] <- "CONTROL" -->
<!-- # sets[sets == "RD-VA-03"] <- "CONTROL" -->
<!-- # sets[sets == "RD-VA-05"] <- "CONTROL" -->
<!-- # sets[sets == "RD-VA-07"] <- "CONTROL" -->
<!-- # sets[sets == "RD-VA-09"] <- "CONTROL" -->
<!-- # sets[sets == "RD-VA-10"] <- "CONTROL" -->
<!-- # data.combined$set = sets -->
<!-- # fact = factor(sets) -->
<!-- # data.combined@active.ident = fact -->
<!-- # ``` -->


# `r colorize('UMAP Plots of 2 Groups for 10 samples', 'blue')` 

```{r,results='asis', echo=FALSE, message=FALSE, warning=FALSE}
DimPlot(data.combined, reduction = "umap", group.by = "group", raster=FALSE,pt.size =0.001)+ ggtitle('2 Groups After Filtering Top 5 Clusters')
```

# `r colorize('Find Markers For 2 Groups', 'blue')`  {.tabset}

All samples are split into two groups "CR" and "CONTROL", where "CR" contains the set of samples {RD-VA-01, RD-VA-08, RD-VA-11, RD-VA-12}(note that we removed RD-VA-04 and RD-VA-06), "CONTROL" contains the set of samples {RD-VA-02, RD-VA-03, RD-VA-05, RD-VA-07, RD-VA-09, RD-VA-10}. We here find markers for these two groups, report only the positive ones. You can check the top 20 markers(selected based on the largest values of avg_log2FC or pct.diff or pct.ratio) and also all markers for each cluster on the following data table. 


```{r, eval=set_eval, echo=FALSE}
# find markers for every cluster compared to all remaining cells, report only the positive ones

DefaultAssay(data.combined) <- "RNA"

data.markers <- FindAllMarkers(data.combined, only.pos = TRUE, min.pct = 0.1, logfc.threshold = 0.1)

data.markers$pct.diff = data.markers$pct.1 - data.markers$pct.2
data.markers$pct.ratio = data.markers$pct.1 / data.markers$pct.2

# Save markers
write.csv(data.markers, file='data_all_markers_for_10_samples_5_celltypes.csv')

```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
# Read Markers .csv
data.markers = read.csv('data_all_markers_for_10_samples_5_celltypes.csv')
avglog = data.markers %>% group_by(cluster) %>% slice_max(n = 20, order_by = avg_log2FC)
pctdiff = data.markers %>% group_by(cluster) %>% slice_max(n = 20, order_by = pct.diff)
pctratio = data.markers %>% group_by(cluster) %>% slice_max(n = 20, order_by = pct.ratio)
```


```{r, eval = FALSE, echo=FALSE, message=FALSE, warning=FALSE}

positive.markers = read.csv('data_all_markers_for_10_samples_5_celltypes.csv')

positive.markers$pct.diff = positive.markers$pct.1 - positive.markers$pct.2
positive.markers$pct.ratio = positive.markers$pct.1 / positive.markers$pct.2

# Keep positive markers only
positive.markers <- positive.markers[positive.markers$avg_log2FC > 0.0,]

positive.markers <- positive.markers %>%
  arrange(p_val) %>%     # Arrange by p_val in ascending order
  group_by(cluster) %>%  # Group by cluster
  slice_head(n = 200000) # Select top 200000 rows per group

write.csv(positive.markers, file='data_positive_markers_for_10_samples_5_celltypes.csv')

```

```{r,results='asis', echo=FALSE, message=FALSE, warning=FALSE}
cat('##',"Top 20 Markers (avg_log2FC)",' \n')
paged_table(avglog)
cat(' \n \n')
cat('##',"Top 20 Markers (pct.diff)",' \n')
paged_table(pctdiff)
cat(' \n \n')
cat('##',"Top 20 Markers (pct.ratio)",' \n')
paged_table(pctratio)
cat(' \n \n')
cat('##',"All Markers for Each Subcluster",' \n')
paged_table(data.markers)
cat(' \n \n')

```



#

The above data frames have the following columns:

* **cluster**: cluster label.
* **gene**: gene name
* **p_val**: p value (unadjusted) can take any value between 0 and 1. Values close to 0 indicate that the observed gene expression difference between the cluster and the rest clusters is unlikely to be due to chance, whereas a p value close to 1 suggests no difference between the cluster and the rest clusters other than due to chance.
* **avg_log2FC**: log fold-change of the average expression between the cluster and the rest clusters. Positive values indicate that the gene is more highly expressed in the cluster.
* **pct.1**: The percentage of cells where the gene is detected in the cluster.
* **pct.2**: The percentage of cells where the gene is detected in the rest clusters.
* **p_val_adj**: Adjusted p value, based on Bonferroni correction using all genes in the dataset.
* **pct.diff**: The difference between pct.1 and pct.2.
* **pct.ratio**: The ratio between pct.1 and pct.2.




# `r colorize('Gene Enrichment', 'blue')`


```{r,results='asis', echo=FALSE, message=FALSE, warning=FALSE}
library("RDAVIDWebService")
library("org.Hs.eg.db")
library("xlsx")

num=250

replacENTREZID <- function(functionMatrix, mapping) {
  for(i in 1:nrow(functionMatrix)) {
    genelist <- trimws(strsplit(functionMatrix[i,]$Genes, ",")[[1]])
    geneName <- mapping[match(genelist, mapping$ENTREZID), 1]
    functionMatrix[i,]$Genes <- paste(geneName, collapse = ",")
  }
  return(functionMatrix)
}

  david <- DAVIDWebService$new(email='saumyapo@bu.edu', url="https://david.ncifcrf.gov/webservice/services/DAVIDWebService.DAVIDWebServiceHttpSoap12Endpoint/")
  
  ########################## L1
  DEG.list <- read.csv("../../02_Block/L1_common_all_pvals.csv", header=T, stringsAsFactors=FALSE, check.names=FALSE)
  DEG.list <- DEG.list[order(DEG.list$edgington),]
  colnames(DEG.list)[1] <- "Symbol"
  top.up.pF <- DEG.list[DEG.list$AvgLogFC>0,]
  top.down.pF <- DEG.list[DEG.list$AvgLogFC<0,]
  
  egids <- AnnotationDbi::select(org.Hs.eg.db, keys=top.up.pF[1:num,1], columns=c("SYMBOL", "ENTREZID"), keytype="SYMBOL")
  result <- addList(david, egids[,2], idType = "ENTREZ_GENE_ID", listName = "L1.UP", listType = "Gene")
  setAnnotationCategories(david, c("GOTERM_BP_DIRECT", "KEGG_PATHWAY", "REACTOME_PATHWAY"))
  func.up.L1 <- getFunctionalAnnotationChart(david)
  func.up.L1 <- replacENTREZID(func.up.L1, egids)
```




```{r,results='asis', echo=FALSE, message=FALSE, warning=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.12")

BiocManager::valid()
```

```{r,results='asis', echo=FALSE, message=FALSE, warning=FALSE}
BiocManager::install(c(
    "adabag", "additivityTests", "ade4", "AER", "affxparser", "affy", "affycoretools", "affyio", "affyPLM",
    "AICcmodavg", "airway", "akima", "AlgDesign", "ALL", "alphavantager", "amap", "Amelia", "annaffy",
    "annotate", "AnnotationDbi", "AnnotationFilter", "AnnotationForge", "AnnotationHub", "antiword",
    "anytime", "apcluster", "ape", "aqp", "argparse", "arm", "aroma.affymetrix", "aroma.apd", "aroma.core",
    "aroma.light", "arrangements", "ArrayExpressHTS", "arules", "arulesViz", "assertive", "assertive.base",
    "assertive.code", "assertive.datetimes", "assertive.properties", "assertive.reflection", "ASSIGN",
    "audio", "available", "aweek", "aws.signature", "backports", "ballgown", "base64", "baseline",
    "BaseSpaceR", "BatchExperiments", "BatchJobs", "BatchQC", "batchtools", "bayesplot", "bayestestR",
    "BBmisc", "bbmle", "BDgraph", "bdsmatrix", "beachmat", "beadarraySNP", "beanplot", "beeswarm", "bench",
    "betareg", "BH", "BHC", "BiasedUrn", "bibtex", "biclust", "biganalytics", "bigassertr", "biglm",
    "bigmemory", "bigmemory.sri", "bigparallelr", "bigreadr", "bigsnpr", "bigsparser", "bigstatsr",
    "bigtabulate", "bigutilsr", "bindata", "Biobase", "BiocCheck", "BiocFileCache", "BiocGenerics",
    "BiocManager", "BiocNeighbors", "BiocParallel", "BiocSingular", "BiocStyle", "BiocVersion", "biocViews",
    "bioDist", "biomaRt", "biomartr", "biomformat", "Biostrings", "biovizBase", "bit", "bit64", "bitops",
    "bladderbatch", "blme", "blob", "blockmodeling", "BMA", "bnlearn", "Bolstad2", "bookdown", "brew",
    "brglm", "bridgedist", "bridgesampling", "Brobdingnag", "broom", "BSgenome", "bsseq", "bumphunter", "C50",
    "Cairo", "calibrate", "callr", "cancerdata", "car", "carData", "Cardinal", "CardinalWorkflows", "caret",
    "Category", "caTools", "cba", "CDM", "celda", "censReg", "ChAMP", "ChAMPdata", "checkmate", "chimera",
    "chopsticks", "chron", "circlize", "classInt", "cli", "clime", "ClimProjDiags", "clipr", "clue", "clv",
    "coarseDataTools", "cobs", "coda", "coin", "coloc", "colorspace", "colourpicker", "colourvalues",
    "commonmark", "ComplexHeatmap", "config", "ConsensusClusterPlus", "coop", "corpcor", "corrgram",
    "corrplot", "covr", "CovTools", "cowplot", "crayon", "CRISPRseek", "crosstalk", "crul", "cubature",
    "Cubist", "cummeRbund", "curl", "CVST", "dada2", "data.table", "dataMaid", "date", "DBI", "dbplyr",
    "dbscan", "dcurver", "ddalpha", "DECIPHER", "DelayedArray", "DelayedMatrixStats", "deldir", "dendextend",
    "densityClust", "DEoptimR", "depmixS4", "Deriv", "desc", "descr", "DescTools", "DESeq2", "deSolve",
    "devtools", "DEXSeq", "dfidx", "dfoptim", "DiagrammeR", "dichromat", "DiffBind", "diffobj", "digest",
    "dimRed", "diptest", "dirmult", "distillery", "distr", "DMRcate", "DMRcatedata", "DNAcopy", "doBy",
    "docopt", "doMC", "doParallel", "doRNG", "DOSE", "doSNOW", "dotCall64", "dplyr", "dqrng", "DropletUtils",
    "DSS", "DT", "dtplyr", "dtw", "e1071", "earlyR", "earth", "EBImage", "EBSeq", "EDASeq", "edgeR",
    "EFAtools", "effects", "effectsize", "ellipse", "ellipsis", "emulator", "energy", "english",
    "EnhancedVolcano", "enrichplot", "enrichR", "ensembldb", "entropy", "epiDisplay", "EpiEstim", "epiR",
    "epitrix", "ergm.count", "esquisse", "estimability", "europepmc", "evaluate", "evd", "ExperimentHub",
    "expint", "expm", "extraDistr", "extrafont", "extRemes", "FactoMineR", "fANCOVA", "fanplot", "fansi",
    "farver", "fastcluster", "fastDummies", "fastGHQuad", "fastICA", "fastmap", "fastmatch", "fauxpas",
    "fBasics", "fda", "fdrtool", "ff", "fftw", "fftwtools", "fgsea", "fields", "findpython", "fission",
    "fit.models", "fitdistrplus", "flexclust", "flexmix", "FNN", "forcats", "foreach", "forecast", "formatR",
    "formattable", "Formula", "fpc", "fracdiff", "fs", "furrr", "future", "future.apply", "GA", "gam",
    "gamlss", "gamlss.data", "gamlss.dist", "gap", "gargle", "gaston", "gcrma", "gdata", "gdsfmt", "gdtools",
    "gee", "geepack", "geiger", "gender", "genefilter", "geneLenDataBase", "geneplotter", "generics",
    "GENESIS", "genetics", "GenomeInfoDb", "GenomeInfoDbData", "GenomicAlignments", "GenomicFeatures",
    "GenomicRanges", "geojsonsf", "geometry", "geomorph", "GEOquery", "geoR", "geosphere", "GetoptLong",
    "GGally", "ggbeeswarm", "ggbio", "ggdendro", "ggeffects", "ggExtra", "ggforce", "ggfortify",
    "gghighlight", "ggmap", "ggmcmc", "ggnetwork", "ggplot2", "ggplotify", "ggpubr", "ggraph", "ggrepel",
    "ggridges", "ggsci", "ggsignif", "ggthemes", "ggvis", "gh", "git2r", "Glimma", "glmmML", "glmmTMB",
    "glmnet", "glmnetUtils", "GlobalOptions", "globals", "globaltest", "glue", "gmailr", "gmm", "GMMAT",
    "gmodels", "gmp", "gnm", "GO.db", "goftest", "googlePolylines", "googleVis", "GOSemSim", "goseq",
    "GOstats", "gower", "GPArotation", "gplots", "graph", "graphite", "graphlayouts", "graphon",
    "gridGraphics", "gridSVG", "grImport", "gsalib", "GSEABase", "gss", "gstat", "GSVA", "GSVAdata", "gt",
    "gtable", "gtools", "Gviz", "GWASTools", "gWidgets2", "h2o", "HaploSim", "hash", "haven", "HDF5Array",
    "heatmap3", "Heatplus", "heplots", "here", "hexbin", "HiddenMarkov", "highcharter", "highlight", "highr",
    "HistData", "Hmisc", "hms", "hoardr", "Hotelling", "HSAUR", "HSMMSingleCell", "htmlTable", "htmltools",
    "htmlwidgets", "httpuv", "httr", "huge", "hunspell", "hwriter", "hyperSpec", "ica", "ICC", "igraph",
    "IHW", "Illumina450ProbeVariants.db", "illuminaio", "impute", "incidence", "influenceR", "infotheo",
    "inline", "insight", "interactiveDisplayBase", "interplot", "intervals", "inum", "ipred", "IRanges",
    "IRdisplay", "irlba", "isa2", "ISLR", "isoband", "ISOcodes", "iterators", "janeaustenr", "janitor",
    "JavaGD", "JGR", "JM", "jomo", "jpeg", "jsonify", "jsonlite", "kableExtra", "KEGGgraph", "KEGGREST",
    "Kendall", "keras", "kernlab", "kinship2", "klaR", "kml", "knitr", "ks", "labdsv", "labeling", "labelled",
    "laeken", "Lahman", "LambertW", "lamW", "LaplacesDemon", "lars", "latentnet", "later", "latticeExtra",
    "lava", "lavaan", "lcmm", "leaflet", "leiden", "lfe", "lhs", "libcoin", "lifecycle", "limma", "linprog",
    "lintr", "listenv", "littler", "lme4", "lmerTest", "lmomco", "lmtest", "logcondens", "logistf", "logitT",
    "logspline", "lokern", "longitudinalData", "loo", "lpSolve", "lpsymphony", "lsei", "lsr", "ltm",
    "lubridate", "lumi", "magic", "magrittr", "MALDIquant", "manipulateWidget", "mapdata", "mapdeck",
    "mapproj", "maps", "maptools", "maptree", "markdown", "marray", "MassSpecWavelet", "MAST", "Matching",
    "MatchIt", "matlab", "matrixcalc", "MatrixModels", "matrixStats", "matter", "maxLik", "mc2d", "mclust",
    "MCMCglmm", "MCMCpack", "memoise", "memuse", "MendelianRandomization", "merTools", "meta", "metafor",
    "metagenomeSeq", "metap", "MetaSKAT", "methylumi", "mgsub", "mi", "mice", "micEcon", "microbenchmark",
    "microRNA", "mime", "minfi", "minpack.lm", "minqa", "mirt", "misc3d", "miscTools", "missMethyl", "mitml",
    "mix", "mixmeta", "mixOmics", "mixtools", "mlbench", "mlegp", "mlogit", "mnormt", "mockery", "modelr",
    "moments", "monocle", "MotifDb", "motifStack", "MPV", "MSGARCH", "msigdbr", "msir", "msm", "mstate",
    "multcomp", "multgee", "multiApply", "multicool", "MultiDataSet", "multitaper", "multtest", "mutoss",
    "mvnfast", "mvnmle", "mvoutlier", "mvtnorm", "mzID", "NbClust", "NBPSeq", "ncbit", "netbiov",
    "NetPathMiner", "network", "NetworkDistance", "networkDynamic", "nimble", "nleqslv", "nloptr", "NLP",
    "NMF", "NMOF", "NOISeq", "nonnest2", "np", "npsurv", "nycflights13", "OceanView", "officer", "oligo",
    "oligoClasses", "openair", "openssl", "openxlsx", "optimx", "optiscale", "optparse", "ordinal",
    "org.Bt.eg.db", "org.Cf.eg.db", "org.Dm.eg.db", "org.Hs.eg.db", "org.Mm.eg.db", "org.Rn.eg.db",
    "org.Ss.eg.db", "OrganismDbi", "orthopolynom", "osmdata", "outliers", "oz", "packrat", "padr", "pals",
    "pan", "pander", "parameters", "ParamHelpers", "parsedate", "partitions", "party", "partykit",
    "patchwork", "pathifier", "PathoStat", "pbapply", "PBSmapping", "pcaMethods", "pcaPP", "PCICt",
    "pedigree", "pedometrics", "penalized", "peperr", "performance", "permute", "PFAM.db", "philentropy",
    "phyloseq", "Pigengene", "pillar", "pinfsc50", "pixmap", "pkgbuild", "pkgKitten", "pkgload", "pkgmaker",
    "PKI", "plm", "plot3D", "plot3Drgl", "plotly", "plotmo", "plotrix", "plotROC", "pls", "plyr", "pmml",
    "png", "polspline", "polyclip", "polycor", "polyester", "polynom", "pracma", "prebs", "preprocessCore",
    "prettydoc", "princurve", "pROC", "processx", "prodlim", "profileModel", "profmem", "proftools",
    "profvis", "promises", "ProtGenerics", "proxy", "proxyC", "pryr", "ps", "PSCBS", "pscl", "pspline",
    "psych", "ptw", "purrr", "purrrlyr", "qap", "qdapRegex", "qdapTools", "qgam", "qgraph", "qpdf", "qqman",
    "qrnn", "qtl", "Quandl", "quanteda", "quantmod", "quantreg", "quantsmooth", "questionr", "qvalue",
    "qvcalc", "qwraps2", "R.cache", "R.devices", "R.filesets", "R.huge", "R.matlab", "R.methodsS3", "R.oo",
    "R.rsp", "R.utils", "r2d3", "R2HTML", "R6", "randtoolbox", "ranger", "RApiDatetime", "rappdirs",
    "rapportools", "raster", "rasterVis", "rattle", "RBGL", "RBioinf", "rBiopaxParser", "rcdk", "rcdklibs",
    "Rcgmin", "rcmdcheck", "RColorBrewer", "Rcpp", "RcppAlgos", "RcppAnnoy", "RcppArmadillo", "RcppCNPy",
    "RcppEigen", "RcppGSL", "RcppHNSW", "RcppParallel", "RcppThread", "RcppZiggurat", "rcrossref", "Rcsdp",
    "RCurl", "Rdpack", "reactome.db", "readODS", "readr", "readstata13", "readxl", "recipes", "RecordLinkage",
    "REddyProc", "relaimpo", "relations", "relevent", "remotes", "rentrez", "renv", "Repitools",
    "ReportingTools", "repr", "reprex", "reshape", "reticulate", "rex", "Rfast", "rGADEM", "RGCCA", "rgenoud",
    "rgl", "RGraphics", "Rgraphviz", "rhdf5", "Rhdf5lib", "RhpcBLASctl", "Rhtslib", "riingo", "Ringo",
    "RInside", "rio", "rJava", "rjson", "RJSONIO", "Rlab", "rlang", "rlecuyer", "rlist", "RMariaDB",
    "rmarkdown", "rmdformats", "rmio", "Rmisc", "Rmpfr", "rms", "RMTstat", "RNAseqData.HNRNPC.bam.chr14",
    "rngtools", "rngWELL", "rnoaa", "robCompositions", "robust", "robustbase", "ROC", "ROI", "Rook", "ropls",
    "ROptSpace", "ROTS", "roxygen2", "RPA", "rpart.plot", "rpf", "RPostgreSQL", "rprojroot", "rrcov", "RRPP",
    "rsample", "Rsamtools", "rsconnect", "Rserve", "RSiena", "RSpectra", "RSQLite", "rstan", "rstanarm",
    "rstantools", "rstatix", "rstpm2", "rstream", "rstudioapi", "Rsubread", "rsvd", "rsvg", "rtracklayer",
    "Rtsne", "Rttf2pt1", "RUVSeq", "rvcheck", "rversions", "rvest", "rvg", "Rvmmin", "S4Vectors",
    "sampleSelection", "sampling", "sandwich", "sas7bdat", "sass", "satellite", "SBMLR", "SC3", "scales",
    "SCAN.UPC", "scater", "scatterplot3d", "scde", "scran", "scRNAseq", "sctransform", "segmented", "sem",
    "semTools", "sendmailR", "SeqArray", "seqinr", "seqLogo", "seqminer", "SeqVarTools", "seriation",
    "sessioninfo", "setRNG", "sets", "settings", "Seurat", "sfheaders", "sfsmisc", "shape", "shapefiles",
    "shapes", "shiny", "shinyalert", "shinyBS", "shinycssloaders", "shinydashboard", "shinyjs", "shinystan",
    "shinythemes", "shinyWidgets", "ShortRead", "SHT", "siggenes", "signal", "SingleCellExperiment", "sirt",
    "sitmo", "sjlabelled", "sjmisc", "sjPlot", "sjstats", "SKAT", "slam", "slider", "sm", "sn", "sna", "snow",
    "SnowballC", "snowfall", "SNPRelate", "snpStats", "sodium", "soilDB", "solartime", "sourcetools", "sp",
    "spacetime", "spam", "sparklyr", "SparseM", "sparsesvd", "SPAtest", "spatialwidget", "spatstat",
    "spatstat.data", "spatstat.utils", "spBayes", "spData", "speedglm", "SPEI", "splancs", "splines2",
    "splus2R", "SQUAREM", "StanHeaders", "stargazer", "statmod", "statnet.common", "stopwords", "stringdist",
    "stringi", "stringr", "striprtf", "strucchange", "styler", "subplex", "SummarizedExperiment", "survey",
    "survRM2", "sva", "svglite", "svMisc", "svUnit", "sys", "systemfit", "systemfonts", "systemPipeR",
    "syuzhet", "table1", "tableone", "TAM", "tclust", "tensorA", "tensorflow", "TENxBrainData", "tergm",
    "testthat", "textshape", "tfplot", "tfruns", "tgp", "TH.data", "tibble", "tictoc", "tidygraph",
    "tidyquant", "tidyr", "tidyselect", "tidytext", "tidytree", "tidyverse", "tiff", "timeDate", "timeSeries",
    "timetk", "tinytex", "tis", "tkrplot", "tm", "TMB", "tmvtnorm", "tokenizers", "treeio", "triebeard",
    "TripleR", "truncnorm", "TSA", "tseries", "tsna", "tsne", "TSP", "TTR", "tvReg", "tweenr",
    "TxDb.Mmulatta.UCSC.rheMac8.refGene", "tximport", "ucminf", "umap", "unmarked", "urca", "uroot",
    "usethis", "UsingR", "utf8", "uuid", "uwot", "V8", "VariantAnnotation", "vars", "vcd", "vcdExtra", "vcfR",
    "vctrs", "vdiffr", "vegan", "VennDiagram", "VGAM", "VIM", "VineCopula", "vioplot", "viridis",
    "viridisLite", "visNetwork", "vsn", "warp", "wateRmelon", "waveslim", "webshot", "WGCNA", "whisker",
    "withr", "Wrench", "WriteXLS", "xfun", "xgboost", "xlsx", "XML", "xml2", "xmlparsedata", "xts", "XVector",
    "yaml", "yesno", "zCompositions", "zebrafishRNASeq", "zip", "zlibbioc", "zoo", "zyp"
  ), update = TRUE, ask = FALSE)
```








